src/libp2p/Libp2p.Core/Peer.cs

```diff
@@ -0,0 +1,660 @@
+                    break;
+                }
+            default:
+                if (isListener && top.Protocol is ISessionListenerProtocol listenerProtocol)
+                {
+                    SessionContext ctx = new(this, session, top, isListener, options);
+                    upgradeTask = listenerProtocol.ListenAsync(parentChannel, ctx);
+                    break;
+                }
+
+                Type? genericInterface = top.Protocol.GetType().GetInterfaces()
+                    .FirstOrDefault(i =>
+                        i.IsGenericType &&
+                        i.GetGenericTypeDefinition() == typeof(ISessionProtocol<,>));
+
+                if (genericInterface != null)
+                {
+                    Type[] genericArguments = genericInterface.GetGenericArguments();
+                    Type requestType = genericArguments[0];
+                    Type responseType = genericArguments[1];
+
+                    if (options?.Argument is not null && !options.Argument.GetType().IsAssignableTo(requestType))
+                    {
+                        throw new ArgumentException($"Invalid request. Argument is of {options.Argument.GetType()} type which is not assignable to {requestType.FullName}");
+                    }
+
+                    // Dynamically invoke DialAsync
+                    System.Reflection.MethodInfo? dialAsyncMethod = genericInterface.GetMethod("DialAsync");
+                    if (dialAsyncMethod != null)
+                    {
+                        SessionContext ctx = new(this, session, top, isListener, options);
+                        upgradeTask = (Task)dialAsyncMethod.Invoke(top.Protocol, [parentChannel, ctx, options?.Argument])!;
+                        break;
+                    }
+                }
+                throw new Libp2pSetupException($"Protocol {top.Protocol} does not implement proper protocol interface");
+        }
+
+        if (options?.SelectedProtocol == top.Protocol && options?.CompletionSource is not null)
+        {
+            _ = upgradeTask.ContinueWith(async t =>
+            {
+                MapToTaskCompletionSource(t, options.CompletionSource);
+                await parentChannel.CloseAsync();
+            });
+        }
+
+        await upgradeTask.ContinueWith(t =>
+        {
+            if (t.IsFaulted)
+            {
+                _logger?.LogError($"Upgrade task failed with {t.Exception}");
+            }
+            _ = parentChannel.CloseAsync();
+            _logger?.LogInformation($"Finished#2 {protocol} to {top}, listen={isListener}");
+        });
+    }
+
+    public Task DisconnectAsync() => Task.WhenAll(sessions.ToArray().Select(s => s.DisconnectAsync()));
+}
+
+public class NewSessionContext(LocalPeer localPeer, LocalPeer.Session session, ProtocolRef protocol, bool isListener, UpgradeOptions? upgradeOptions) : ContextBase(localPeer, session, protocol, isListener, upgradeOptions), INewSessionContext
```

Comment:
Separate classes into their own files. Maybe break up the `Peer` class if possible as this is a big file